File: kotlin/012_RedPainting.kt
/**
 * 012 - Red Painting（★4）
 * Union-Find
 */

/**
 * 012 - Red Painting（★4）
 * Union-Find
 */

import java.io.PrintWriter

class UnionFind(val size: Int) {
  val parents = IntArray(size) { it }

  fun root(u: Int): Int {
    if (parents[u] == u) return u
    parents[u] = root(parents[u])
    return parents[u]
  }

  fun unite(u: Int, v: Int) {
    val ru = root(u)
    val rv = root(v)
    if (ru == rv) return
    parents[ru] = rv
  }

  fun isSame(u: Int, v: Int): Boolean {
    return root(u) == root(v)
  }
}

fun main() {
  val DIRECT = listOf(1 to 0, 0 to 1, -1 to 0, 0 to -1)

  val (H, W) = readLine()!!.split(" ").map(String::toInt)
  val Q = readLine()!!.toInt()
  val QUERY = Array(Q) { readLine()!!.split(" ").map(String::toInt) }

  val red = Array(H) { BooleanArray(W) { false } }
  val uf = UnionFind(H * W)

  val pw = PrintWriter(System.out, false)

  QUERY.forEach { q ->
    when (q[0]) {
      1 -> {
        val (r, c) = q.subList(1, 3).map { it - 1 }
        red[r][c] = true
        DIRECT.forEach direct@ { (dr, dc) ->
          val nr = r + dr
          val nc = c + dc
          if (nr !in 0 until H || nc !in 0 until W || !red[nr][nc]) return@direct
          uf.unite(r * W + c, nr * W + nc)
        }
      }
      2 -> {
        val (ra, ca, rb, cb) = q.subList(1, 5).map { it - 1 }
        var ans = "No"
        if (red[ra][ca] && red[rb][cb] && uf.isSame(ra * W + ca, rb * W + cb)) ans = "Yes"
        pw.println(ans)
      }
    }
  }
  pw.flush()
  pw.close()
}

File: ruby/012_RedPainting.rb
# 012 - Red Painting（★4）
# Union-Find

class UnionFind
  attr_accessor :parents #, :relations

  def initialize(n)
    @parents = Array.new(n, &:itself)
    # @relations = {}
    # n.times { @relations[_1] = [_1] }
  end

  def root(u)
    return u if u == parents[u]
    parents[u] = root(parents[u])
  end

  def same?(u, v)
    root(u) == root(v)
  end

  def unite(u, v)
    ru, rv = root(u), root(v)
    return if ru == rv
    # relations[_v] |= relations[_u]
    parents[ru] = rv
  end
end

H, W = gets.split.map(&:to_i)
Q = gets.to_i
QUERY = Array.new(Q) { gets.split.map(&:to_i) }
@red = Array.new(H) { [false] * W }
@uf = UnionFind.new(H * W)

def unite_red(r, c)
  r, c = r - 1, c - 1
  @red[r][c] = true
  [[1,0], [0,1], [-1,0], [0,-1]].each do |dr, dc|
    nr, nc = r + dr, c + dc
    next if !nr.between?(0, H - 1) || !nc.between?(0, W - 1)
    next unless @red[nr][nc]
    @uf.unite(r * W + c, nr * W + nc)
  end
end

def same_red?(ra, ca, rb, cb)
  ra, ca = ra - 1, ca - 1
  rb, cb = rb - 1, cb - 1
  return false unless @red[ra][ca] && @red[rb][cb]
  @uf.same?(ra * W + ca, rb * W + cb)
end

ans = []
QUERY.each do |q|
  if q[0] == 1
    unite_red(*q[1, 2])
  else
    ans << (same_red?(*q[1, 4]) ? "Yes" : "No")
  end
end

puts ans
